treewide: use non-binary wrapper for wrapGAppsHook where appropriate

xaverdh opened the pull request:
The use of --add-flags in this derivation does not work with binary
wrappers, which wrapGAppsHook uses since #164163.

###### Description of changes

See discussion [here](https://github.com/NixOS/nixpkgs/pull/172315) 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

xaverdh committed:
signal-desktop: use non-binary wrapper for wrapGAppsHook

The use of --add-flags in this derivation does not work with binary
wrappers, which wrapGAppsHook uses since #164163.

xaverdh commented:
Ok, the test passed and I am successfully running signal from this branch on my system.

ncfavier commented:
I think we can fix the other programs in this PR as well

xaverdh commented:
> I think we can fix the other programs in this PR as well

will do that!

xaverdh committed:
brave: use non-binary wrapper for wrapGAppsHook

The use of --add-flags in this derivation does not work with binary
    wrappers, which wrapGAppsHook uses since #164163.

xaverdh committed:
discord: use non-binary wrapper for wrapGAppsHook

The use of --add-flags in this derivation does not work with binary
    wrappers, which wrapGAppsHook uses since #164163.

xaverdh committed:
vscode: use non-binary wrapper for wrapGAppsHook

The use of --add-flags in this derivation does not work with binary
    wrappers, which wrapGAppsHook uses since #164163.

xaverdh commented:
I did build `brave`, but not `vscode` and `discord` (due to the unfree license). Still I think they should work.

From a quick grep through nixpkgs, these are indeed all derivations that use `wrapGAppsHook`and mention `NIXOS_OZONE_WL`. But there are probably other similar usages of `--add-flags`..

ncfavier commented:
Yes, but none of those I could find are using `wrapGAppsHook`, so the fix is less urgent.

ncfavier commented:
Please change the PR's title

xaverdh commented:
> Yes, but none of those I could find (`grep -e "--(add-flags|prefix|suffix|set|argv0).*\\\\('')?\\\$"`) are using `wrapGAppsHook`, so the fix is less urgent.

I think e.g. [cryptomator](https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/security/cryptomator/default.nix) is also broken, unless I am mistaken?

ncfavier commented:
It doesn't look like it expands any variables at runtime.

xaverdh commented:
> 

Will the `~` get expanded by the java runtime itself?

ncfavier commented:
At least it won't get expanded by Bash, because of the single quotes.

ncfavier commented:
But [this](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/science/logic/tlaplus/toolbox.nix#L66) will.

xaverdh commented:
> 

Right, still there appears to be an issue with the ``
--add-flags "--class-path '$out/share/cryptomator/libs/*'"
``
part. I get `Error: --class-path requires class path specification`on master but it works when I inject the non-binary wrapper.

ncfavier commented:
Ah, I guess the binary implementation of `--add-flags` doesn't parse quotes.

xaverdh commented:
There are also cases like [this](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/window-managers/sway/wrapper.nix) where the implementation is "leaked" to users, but Its not very likely somebody relies on injecting shell code there?

ncfavier commented:
It looks like we also need to fix `gmtp` (probably just remove quotes around `$out/share`) and `bluej`.

xaverdh commented:
> 

`gmtp` appears to be fine, the $out does get expanded with the binary wrapper..

ncfavier commented:
Yes but the quotes aren't parsed, same issue: `Failed to open file “"/nix/store/myy2l41c99crnp29kvhxl9n6j02qfwni-gmtp-1.3.11/share"/../share/gmtp/gmtpicon.png”: No such file or directory`

xaverdh commented:
> 

ah right. So the proper fix in this cases appears to be to remove the quotes.

xaverdh committed:
tlaplusToolbox: use non-binary wrapper for wrapGAppsHook

The use of --add-flags in this derivation does not work with binary
    wrappers, which wrapGAppsHook uses since #164163.

xaverdh committed:
cryptomator: use non-binary wrapper for wrapGAppsHook

The use of --add-flags in this derivation does not work with binary
    wrappers, which wrapGAppsHook uses since #164163.

xaverdh committed:
bluej: fix with non-binary wrapper for wrapGAppsHook

The use of --add-flags in this derivation assumed quotes to be expanded, which the binary
         wrapper (which wrapGAppsHook uses since #164163) will not do.

xaverdh committed:
gmtp: fix with non-binary wrapper for wrapGAppsHook

The use of --add-flags in this derivation assumed quotes to be expanded, which the binary
         wrapper (which wrapGAppsHook uses since #164163) will not do.

xaverdh commented:
So from a quick look, these appear to be all issues in derivations with both `--add-flags` and `wrapGAppsHook`.

xaverdh commented:
I will take a break now (few hours). Feel free to merge this / push to this pr branch.

thiagokokada commented on pkgs/applications/editors/vscode/generic.nix:
```diff
@@ -68,7 +68,12 @@ let
+      ++ lib.optionals stdenv.isLinux [
+        autoPatchelfHook
+        nodePackages.asar
+        (wrapGAppsHook.override { makeBinaryWrapper = makeWrapper; })
```
This kinda of looks the contrary of what I expect. The argument for `wrapGAppsHook` should be named `makeWrapper`, then set `makeWrapper = makeBinaryWrapper` in `all-packages.nix` and the override it here with `{ inherit makeWrapper; }`  instead.

ncfavier replied:
I agree. I'm making a big "clean up" PR for `makeBinaryWrapper`, I'll include that.

thiagokokada replied:
I am going to say that this is a minor nitpick though. I am not against merging this as-is. It just looks strange.

ncfavier replied:
https://github.com/NixOS/nixpkgs/pull/172366

thiagokokada commented:
BTW, from what I understood from a quick look at the PR the issue is either glob or `~` expansion right?

Could we eventually use those instead?
- https://man7.org/linux/man-pages/man3/glob.3.html
- https://man7.org/linux/man-pages/man3/wordexp.3.html

Not for this PR of course, just for future improvements of `makeBinaryWrapper`.

CC @bergkvist.

thiagokokada commented:
> No glob issues, just variable expansion and tilde expansion. Using `wordexp` would be a good idea, but it doesn't cover parameter expansions like `"\''${NIXOS_OZONE_WL:+\''${WAYLAND_DISPLAY:+--enable-features=UseOzonePlatform --ozone-platform=wayland}}"`. I'm not sure what the best way would be to provide the same flexibility as shell wrappers in binary wrappers.

Huh... I think the easiest way would be to have something like (pseudo-code) `output = popen("sh -c "echo \''${NIXOS_OZONE_WL:+\''${WAYLAND_DISPLAY:+--enable-features=UseOzonePlatform --ozone-platform=wayland}}", "w");` and the passing the result to the `exec`.

Yeah, this is ugly and it is kinda against the original idea of the `makeBinaryWrapper`, but maybe we could have this behind a flag (e.g.: `parse_in_shell`) so we can still avoid using shell in the majority of the cases.

Ideally we would parse those environment variables in someway, however this can be very trick since shell variables are really flexible. We can still implement a small subset, or even maybe a small DSL for those cases.

ncfavier commented on pkgs/applications/science/logic/tlaplus/toolbox.nix:
```diff
@@ -34,7 +34,10 @@ stdenv.mkDerivation rec {
-  nativeBuildInputs = [ makeWrapper wrapGAppsHook ];
+  nativeBuildInputs = [
+    makeWrapper
+    (wrapGAppsHook.override { makeBinaryWrapper = makeWrapper; }
```
```suggestion
    (wrapGAppsHook.override { makeBinaryWrapper = makeWrapper; })
```

xaverdh replied:
Done (contrary to what I said above..)

drupol commented:
Do you think we can commit this as-is so we unblock people using vscode and other software impacted by this issue ?

thiagokokada commented:
Done.

drupol commented:
Thanks ! Any clue on how to force a new evaluation on hydra?

thiagokokada commented:
> Thanks ! Any clue on how to force a new evaluation on hydra?

No, but not sure if this is a good idea. I mean, almost every PR here is important for those that need it, if we had to restart hydra evals on each PR merged we would never get anywhere.

ncfavier commented:
Started a new evaluation. I think vscode + signal + discord is past the threshold of "important for everyone".

drupol commented:
Thank you :)

dotlambda commented on pkgs/tools/security/cryptomator/default.nix:
```diff
@@ -98,7 +98,13 @@ in stdenv.mkDerivation rec {
+    autoPatchelfHook
+    maven
+    makeWrapper
+    (wrapGAppsHook.override { makeBinaryWrapper = makeWrapper; })
```
Why is the override needed for cryptomator?

ncfavier replied:
https://github.com/NixOS/nixpkgs/pull/172320#issuecomment-1122288164

We could also remove the single quotes, but then we would be implicitly relying on the binary wrapper being in scope, which isn't great.

dotlambda replied:
So makeBinaryWrapper end up in cryptomator's nativeBuildInputs without? That should not happen.

ncfavier replied:
> So makeBinaryWrapper ends up in cryptomator's nativeBuildInputs without? That should not happen.

Yes, because it's a propagated build input of the wrapGApps hook (that is how setup hooks work). I agree that's not great, but I don't know how we should fix it. This is a limitation of Bash more than anything.

> The derivation doesn't use gappsWrapperArgs.

It doesn't need to. It first creates the `cryptomator` executable as a wrapper over java, and then wrapGAppsHook wraps that wrapper automatically in the fixup phase.

dotlambda replied:
> Yes, because it's a propagated build input of the wrapGAppsHook (that is how setup hooks work). I agree that's not great, but I don't know how we should fix it. This is a limitation of Bash more than anything.

Just use an absolute path to makeBinaryWrapper in the hook and don't propagate it.

ncfavier replied:
makeBinaryWrapper is also a setup hook, i.e. a Bash script that defines functions. Functions aren't scoped in Bash, so we would need to do something like

```bash
wrapGApp() (
    local program="$1"
    shift 1
    source @makeWrapper@
    wrapProgram "$program" "${gappsWrapperArgs[@]}" "$@"
)
```

in order not to expose the functions to the global scope. Maybe it could work.

ncfavier replied:
But makeWrapper also depends on dieHook, so we'd also need to source that explicitly... it's kind of awkward.

dotlambda replied:
We could call one of the bash functions makeBinaryWrapper and use a boolean argument for wrapGAppsHook

dotlambda replied:
Then let's call the other makeShellWrapper so people can decide which one they want to use if they feel like it. That would make this override superfluous.

bergkvist commented:
@thiagokokada - wordexp looks interesting. As for patterns like `${VAR:-VALUE}` and `${VAR:+VALUE}`, it would be nice if we had some replacement for these in the binary wrappers.

I think
```sh
--set X '${Y:-$Z}'
```
would be equivalent to this (assuming wordexp is able to expand '$Y' and '$Z' at runtime)
```sh
--set X '$Y'
--set-default X '$Z'
```

Need to think a bit more about `${VAR:+VALUE}` though. It might make sense to introduce some new "primitive" to make something equivalent to this possible in binary wrappers. I'm also curious about the performance impact of using wordexp.

thiagokokada commented:
> @thiagokokada - wordexp looks interesting. As for patterns like `${VAR:-VALUE}` and `${VAR:+VALUE}`, it would be nice if we had some replacement for these in the binary wrappers.
> 
> I think
> 
> ```shell
> --set X '${Y:-$Z}'
> ```
> 
> would be equivalent to this (assuming wordexp is able to expand '$Y' and '$Z' at runtime)
> 
> ```shell
> --set X '$Y'
> --set-default X '$Z'
> ```
> 
> Need to think a bit more about `${VAR:+VALUE}` though. It might make sense to introduce some new "primitive" to make something equivalent to this possible in binary wrappers. I'm also curious about the performance impact of using wordexp.

Huh... Actually re-reading wordexp documentation it really looks like what we want (it expand env vars too, I thought it only expanded things like `~`). So yeah, looks exactly what we want. Needs more testing though.

@bergkvist Can you open an issue for this? I think it would be useful to concentrate the discussion there, instead of a closed PR.
